Adding the StringEncoder transformer

rcap107 opened the pull request:
This is a first draft of a PR to address #1121

I looked at GapEncoder to figure out what to do. This is a very early version just to have an idea of the kind of code that's needed. 

Things left to do:
- [x] Testing
- [ ] Parameter checking?
- [x] Default value for the PCA? 
- [ ] Docstrings
- [ ] Deciding name of the features

rcap107 commented:
Tests fail on minimum requirements because I am using PCA rather than TruncatedSVD for the decomposition, and that raises issues with potentially sparse matrices.

@jeromedockes suggests using directly TruncatedSVD to begin with, rather than adding a check on the version. 

Also, I am using tf-idf as vectorizer, should I use something else? Maybe HashVectorizer? 

(writing this down so I don't forget)

GaelVaroquaux commented:
I'm very happy to see this progressing.

Can you benchmark it on the experiments from Leo's paper: this is important for modeling choices (eg the hyper-parameters)

GaelVaroquaux commented:
Actually, let's keep it simple, and use the CARTE datasets, they are good enough: https://huggingface.co/datasets/inria-soda/carte-benchmark

You probably want to instanciate a pipeline that uses TableVectorizer + HistGradientBoosting, but embeds one of the string columns with the StringEncoder (the one that is either higest cardinality, or most "diverse entry" in the sense of https://arxiv.org/abs/2312.09634

rcap107 commented:
> Should we also add this to the text encoder example, along the TextEncoder, MinHashEncoder and GapEncoder? It shows a tiny benchmark on the toxicity dataset.

It's already there, and it shows that StringEncoder has performance similar to that of GapEncoder and runtime similar to that of MinHashEncoder

Vincent-Maladiere commented:
That's very interesting!

jeromedockes commented on CHANGES.rst:
```diff
@@ -14,7 +14,9 @@ It is currently undergoing fast development and backward compatibility is not en
 ------------
-
+* The :class:`StringEncoder` encodes strings using tf-idf and truncated SVD
+  decomposition and provides a cheaper alternative to :class:`TextEncoder`.
```
or :class:`GapEncoder`

jeromedockes commented on example_string_encoder.py:
```diff
@@ -0,0 +1,33 @@
+# %% test string encoder
+import polars as pl
```
I guess this should go in a test file?

rcap107 replied:
ops, I forgot to delete this

jeromedockes commented on examples/02_text_with_string_encoders.py:
```diff
@@ -221,6 +224,25 @@ def plot_box_results(named_results):
 
+# %%
+# |TextEncoder| embeddings are very strong, but they are also quite expensive to
+# train. A simpler, faster alternative for encoding strings is the |StringEncoder|,
```
to use -- we don't actually train them

jeromedockes commented on examples/02_text_with_string_encoders.py:
```diff
@@ -221,6 +224,25 @@ def plot_box_results(named_results):
+# %%
+# |TextEncoder| embeddings are very strong, but they are also quite expensive to
+# train. A simpler, faster alternative for encoding strings is the |StringEncoder|,
+# which works by first performing a tf-idf vectorization of the text, and then
```
maybe tf-idf (computing vectors of rescaled word counts + wikipedia link)

jeromedockes commented on skrub/_string_encoder.py:
```diff
@@ -0,0 +1,132 @@
+        del y
+        self.pipe = Pipeline(
+            [
+                ("tfidf", TfidfVectorizer()),
```
I think @GaelVaroquaux suggested using a HashingVectorizer instead of tfidfvectorizer (i don't think this would require changes elswhere in your code)

jeromedockes commented on skrub/_string_encoder.py:
```diff
@@ -0,0 +1,132 @@
+    def __init__(self, n_components=30):
+        self.n_components = n_components
+
+    def _transform(self, X):
```
maybe this could be inlined into `transform` rather than having a separate function unless I'm missing where `_transform` is called outside of `transform`

jeromedockes commented on skrub/_string_encoder.py:
```diff
@@ -0,0 +1,132 @@
+
+        self._is_fitted = True
+
+        return self.transform(X)
```
I think we could take advantage of `pipe.fit_transform` to avoid tokenizing the input twice during fit

rcap107 replied:
I modified the code, but I am not sure the implementation is what you meant

jeromedockes replied:
yep that's exactly what I meant thanks!

jeromedockes reviewed (commented):
nice! :tada:  looking really good @rcap107  . it will be awesome to have this as the cost of the GapEncoder is the main pain point of the tablevectorizer ATM :rocket: 

I left a couple of minor comments, the main one is just a possible performance improvement if we use the vectorizer's `fit_transform` to avoid tokenizing the input twice during fit

the code is very close to ready and we have an example so once that's done what are the next steps? maybe some lightweight comparisons on a couple of datasets, then merge, then more extensive experiments to decide if it can be come the tablevectorizer's default?

rcap107 commented:
> nice! 🎉 looking really good @rcap107 . it will be awesome to have this as the cost of the GapEncoder is the main pain point of the tablevectorizer ATM 🚀
>
> I left a couple of minor comments, the main one is just a possible performance improvement if we use the vectorizer's `fit_transform` to avoid tokenizing the input twice during fit

I implemented a change, I'm not sure if it's what you meant..

> the code is very close to ready and we have an example so once that's done what are the next steps? maybe some lightweight comparisons on a couple of datasets, then merge, then more extensive experiments to decide if it can be come the tablevectorizer's default?

Sounds good

Vincent-Maladiere commented:
> the code is very close to ready and we have an example so once that's done what are the next steps? maybe some lightweight comparisons on a couple of datasets, then merge, then more extensive experiments to decide if it can be come the tablevectorizer's default?

I like this plan as well!

GaelVaroquaux commented:
Before we merge, I would love a bit of experimenting, with a fairly systematic eye on the choice of parameters.

For instance: HashingVectorizer vs TfIdfVectorizer? If we go for hashingVectorizer, should it be followed by a TfIdfTransformer or not? For the vectorizer, what should be the default values for the analyzer, the n_features and the n_gram range. I suspect that we want 'analyzer="char_wb"' and ngram_range=(3, 4) based on my experience.

These choices are actually very important, and we should drive them using systematic experimentation. I know that it is a lot of work, but it makes a huge difference.

rcap107 commented:
So I went and ran a few experiments and I have some preliminary results. I reused some of the code from the main example, so the plots are not very good. 

I modified the StringEncoder to have
- `vectorizer` to choose between `tfidf` and `hashing` (mapped to `TfidfVectorizer` and `HashingVectorizer`). Default is `tfidf`
- `ngram_range`, which can be any pair of integers >= `(1,1`). Default is `(1,1)`.
- `tf_idf_followup`, which is a boolean flag. If active, the `HashingVectorizer` is followed by a `TfidfTransformer`.
- `analyzer`, which can be `word`, `char`, or `char_wb`. Default is `word`.
I also added:
- `n_features`, which is a parameter for `tfidf`.
- `max_features`, which is a parameter for `hashing`.

Then I tested these parameters:
```python
configurations = {
    "ngram_range": [(1, 1), (3, 4)],
    "analyzer": ["word", "char", "char_wb"],
    "vectorizer": ["hashing", "tfidf"],
    "n_components": [30],
    "tf_idf_followup": [True, False],
}
```


Only tfidf
![](https://github.com/user-attachments/assets/4684d395-9c15-4b1d-b6f8-13b89ec899b5)

Only hashingvectorizer, no tfidf followup
![image](https://github.com/user-attachments/assets/0a406c0c-01a6-4282-8c99-41d42487140b)

Hashingvectorizer with tfidf followup
![image](https://github.com/user-attachments/assets/60ea1d21-5bd3-4f05-924c-cef8758459cc)


Summary is:
- tfidf is faster
- hashing + tfidf has the same prediction performance as tfidf, but it's much slower
- hashing alone is worse than tfidf and is much slower

jeromedockes commented:
thanks for those experiments! I'm surprised the (1, 1 ) ngram with "char" or "char_wb" can perform so well -- isn't it just counting individual characters?

GaelVaroquaux commented:
> I'm surprised the (1, 1 ) ngram with "char" or "char_wb" can perform so well -- isn't it just counting individual characters?

char_wb makes sense: it's counting words and characters

rcap107 commented:
I ran a few more experiments, because the results looked suspicious, and I think what I have now makes much more sense.

Pareto plots (time is on a log scale):
![image](https://github.com/user-attachments/assets/f6d4bcf5-c9a5-4464-8d1b-941695b1890a)

Boxplots, test score on y-axis
![image](https://github.com/user-attachments/assets/12cf71fc-a9a5-4b23-971c-c85e5122fdb3)

Boxplots, time on y-axis
![image](https://github.com/user-attachments/assets/0cb05e1a-74c7-4e38-aaf3-b6d85c60aa49)

I don't know why the previous results were so off, I might have been reusing results across runs :thinking: 

What is consistent is that tf-idf is way faster and has similar performance if not better than hashing vectorizer.

GaelVaroquaux commented:
OK, those results look great. Please do keep the associated data, as one day we may publish a paper on skrub :grin: 

If I can make an editorial decision: let's use the char_wb, ngram_range=(3, 4). It matches what we do in the GapEncoder.

rcap107 commented:
Great, then I'll clean up the script I am using and I'll put it somewhere, so I can find it if we decide to run more in-depth experiments (like by testing more tables, or changing the number of components or idk).

As for the StringEncoder, do I keep the HashingVectorizer at all? Or do I keep only the arguments relative to tf-idf?

jeromedockes commented:
>> I'm surprised the (1, 1 ) ngram with "char" or "char_wb" can perform so well -- isn't it just counting individual characters?
>
> char_wb makes sense: it's counting words and characters

IIUC correctly char_wb prevents char ngrams from crossing word boundaries but they're still only character ngrams no?

rcap107 commented on skrub/tests/test_string_encoder.py:
```diff
@@ -0,0 +1,142 @@
+    df_module.assert_frame_equal(check_df, result)
+
+
+def test_hashing(encode_column, df_module):
```
I think this test is failing because the hashing vectorizer is not deterministic, if that's the case, I'm not sure how to handle it in the test other than exposing RNG everywhere. 

I didn't bother because we haven't decided whether to keep the hashing vectorizer in the first place.

jeromedockes replied:
otherwise in the test you can just check the type and shape of the output and the content of `self.pipe`

rcap107 commented:
For the record, this is the same example with the new defaults 
![image](https://github.com/user-attachments/assets/cd4971e1-991b-472a-921d-34af3711fcff)

![image](https://github.com/user-attachments/assets/d42901d4-473c-4043-8ac6-ffcabaa2a606)

Pretty good result I think

jeromedockes commented:
nice!! do you mind trying the employee_salaries dataset too, as it has columns that are more like categories rather than text?

rcap107 commented:
> nice!! do you mind trying the employee_salaries dataset too, as it has columns that are more like categories rather than text?

I used R2 as metric, and did not test HashingVectorizer. It seems like when columns are more categories than text, then the GapEncoder is doing a better job than all the alternatives.
![image](https://github.com/user-attachments/assets/39096acb-b3da-4f6f-bbcd-dd1857f3ebfb)

![image](https://github.com/user-attachments/assets/8d7235e9-6828-4cf4-aac0-d456d5b36db4)

![image](https://github.com/user-attachments/assets/5d9bac0f-6397-430c-a64a-61de5a57fb82)

Still, StringEncoder is much faster and the performance is pretty good. TextEncoder is not doing quite as well on the other hand. 

Bottom line is, for more "categorical-looking" attributes, use the GapEncoder for better performance, or StringEncoder for decent performance and faster fit time, for text use TextEncoder for better performance or StringEncoder for decent performance and much faster fit time. 

This is from a pretty quick and dirty set of experiments, we could also try something like the heuristics from Leo's paper to choose, but that's going to get way more complicated. 

By the way, where should I put all the code I used for these results? So far I have it in my scratch folder.

rcap107 commented on skrub/_string_encoder.py:
```diff
@@ -0,0 +1,201 @@
+        """
+        del y
+
+        # ERROR CHECKING
+        if self.analyzer not in ["char_wb", "char", "word"]:
+            raise ValueError(f"Unknown analyzer {self.analyzer}")
+
+        if not all(isinstance(x, int) and x > 0 for x in self.ngram_range):
+            raise ValueError(
+                "Values in `ngram_range` must be positive integers, "
+                f"found {self.ngram_range} instead."
+            )
+        if not len(self.ngram_range) == 2:
+            raise ValueError(
+                f"`ngram_range` must have length 2, found {len(self.ngram_range)}."
+            )
+
+        if not isinstance(self.n_components, int) and self.n_components > 0:
+            raise ValueError(
+                f"`n_components` must be a positive integer, found {self.n_components}"
+            )
```
I left in only the first check, the other constraints are in there, but I think they're less strict

GaelVaroquaux commented:
> IIUC correctly char_wb prevents char ngrams from crossing word boundaries but they're still only character ngrams no?

Good point. I was confusing with the "add_word" strategy of the GapEncoder ( https://github.com/skrub-data/skrub/blob/8a542bbe3674527feed8d4b0194dbedfe62ace20/skrub/_gap_encoder.py#L236 ). I would be interested if we could also explore this option. I seem to remember that it can help markedly, though come at a cost

GaelVaroquaux commented:
One last thing (I always come up with more :D ):

we'll have to be very careful to summarize the tradeoffs between the different encoders in a few lines (a few lines, something short and clear :D ) at the top of the corresponding section of the docs. It is very important that we convey to the user what we have learned

rcap107 commented:
> One last thing (I always come up with more :D ):
> 
> we'll have to be very careful to summarize the tradeoffs between the different encoders in a few lines (a few lines, something short and clear :D ) at the top of the corresponding section of the docs. It is very important that we convey to the user what we have learned

This is something for a separate PR though

jeromedockes commented on examples/02_text_with_string_encoders.py:
```diff
@@ -132,7 +135,7 @@ def plot_gap_feature_importance(X_trans):
 # need to find the optimal value for this hyperparameter using either |GridSearchCV|
 # or |RandomizedSearchCV|. We skip this part to keep the computation time for this
-# example small.
+# small example.
```
to keep the computation time <missing word> for this ...

jeromedockes commented on skrub/_string_encoder.py:
```diff
@@ -0,0 +1,186 @@
+    First, apply a tf-idf vectorization of the text, then reduce the dimensionality
+    with a truncated SVD decomposition with the given number of parameters.
+
+    New features will be named `{col_name}_{component}` if the series has a name,
```
I think you need double backticks

jeromedockes commented on skrub/_string_encoder.py:
```diff
@@ -0,0 +1,186 @@
+    Parameters
+    ----------
+    n_components : int, default=30
+        Number of components to be used for the PCA decomposition. Must be a
```
to keep the number of acronyms under control maybe we should stick to "SVD" not "PCA". also we could have the expanded acronyms in parentheses the first time we mention them and links to their wikipedia pages in  a Notes section

jeromedockes commented on skrub/_string_encoder.py:
```diff
@@ -0,0 +1,186 @@
+        Number of components to be used for the PCA decomposition. Must be a
+        positive integer.
+    vectorizer : str, "tfidf" or "hashing"
+        Vectorizer to apply to the strings, either `tfidf` or `hashing` for
```
also here not sure what was your desired formatting -- single backticks will be italic, double for monospace

jeromedockes commented on skrub/_string_encoder.py:
```diff
@@ -0,0 +1,186 @@
+        scikit-learn TfidfVectorizer or HashingVectorizer respectively.
+
+    ngram_range : tuple of (int, int) pairs, default=(3,4)
+        Whether the feature should be made of word or character n-grams.
```
looks like the docs for n_gram_range and analyzer got swapped

jeromedockes commented on skrub/_string_encoder.py:
```diff
@@ -0,0 +1,186 @@
+    analyzer : str, "char", "word" or "char_wb", default="char_wb"
+        The lower and upper boundary of the range of n-values for different
+        n-grams to be extracted. All values of n such that min_n <= n <= max_n
+        will be used. For example an `ngram_range` of `(1, 1)` means only unigrams,
```
same comment about rst vs markdown

jeromedockes commented on skrub/_string_encoder.py:
```diff
@@ -0,0 +1,186 @@
+        Encode string columns as a numeric array with the minhash method.
+    GapEncoder :
+        Encode string columns by constructing latent topics.
+    SimilarityEncoder :
+        Encode string columns as a numeric array with n-gram string similarity.
```
```suggestion
```

jeromedockes commented on skrub/_string_encoder.py:
```diff
@@ -0,0 +1,186 @@
+                            ngram_range=self.ngram_range, analyzer=self.analyzer
+                        ),
+                    ),
+                    ("tsvd", TruncatedSVD(n_components=self.n_components)),
```
as in the textencoder, I think we need to handle the case where we end up with the smaller dimension of the tfidf < self.n_components (could happen for example if fitting on a column with few unique words and setting a large n_components and using the word analyzer). in that case we can do the same as textencoder ie keep `tfidf[:, :self.n_components]`

jeromedockes replied:
(adding that logic might require you to move the svd out of the pipeline)

jeromedockes reviewed (commented):
mostly corner cases remaining :tada:  :)

rcap107 commented:
I tested TableVectorizer with `drop` for `high_cardinality` and the result is pretty bad. I also tested GapEncoder with `add_words=True`, but it does not seem to help here.

![image](https://github.com/user-attachments/assets/2985b431-802b-4e8d-a5ab-a36ec07cb998)

![image](https://github.com/user-attachments/assets/8ab691dd-01e4-44c7-8350-6fb98b63f136)

![image](https://github.com/user-attachments/assets/e108d578-493b-4711-96e3-bfbb72ba56d2)

I also tried to use OrdinalEncoder as @Vincent-Maladiere suggested, but I was not able to run the experiments :thinking: 
```python
from sklearn.preprocessing import OrdinalEncoder

ordinal_encoder = OrdinalEncoder(handle_unknown="use_encoded_value")

ordinal_encoder_pipe = clone(gap_pipe).set_params(
    **{"tablevectorizer__high_cardinality": ordinal_encoder}
)
ordinal_encoder_results = cross_validate(ordinal_encoder_pipe, X, y, scoring="r2")
results.append(("OrdinalEncoder", ordinal_encoder_results))
```
This gives me: 
```sh
    res = func(*args, **kwargs)
  File "/home/rcappuzz/Projects/skrub/skrub/_on_each_column.py", line 547, in _fit_transform_column
    raise ValueError(
ValueError: Transformer OrdinalEncoder.fit_transform failed on column 'division'. See above for the full traceback.
```
Plus a very long traceback I couldn't figure out.

Vincent-Maladiere commented:
Nice! So what is the conclusion regarding the `StringEncoder(1, 1)`? How can it perform so well against `drop` and `OrdinalEncoder`, when it only considers individual characters?

Vincent-Maladiere commented:
I'm happy that the string encoder looks like a great baseline for short, messy columns and long, free-form text as well.

rcap107 commented:
> we'll have to be very careful to summarize the tradeoffs between the different encoders in a few lines (a few lines, something short and clear :D ) at the top of the corresponding section of the docs. It is very important that we convey to the user what we have learned This is something for a separate PR though
> I'd rather not. IMHO the docs need to be reorganized as we add complexity to the package. Also, the evidence for this recommendation comes from this PR.

I updated the doc page on the Encoders, but it was only to add the `StringEncoder` and a short summary of the different methods. Looking at the page, I think it would be better to expand on it with more detail for all encoders and maybe an explanation of the parameters, but that's something that would take way more effort (and definitely something for a separate PR).

rcap107 commented:
> Nice! So what is the conclusion regarding the `StringEncoder(1, 1)`? How can it perform so well against `drop` and `OrdinalEncoder`, when it only considers individual characters?

![image](https://github.com/user-attachments/assets/7ca049cb-91fc-4518-891a-5565c1f02f7b)

My feeling is that OrdinalEncoder is just not that good if there is no order in the feature to begin with, while strings that are similar to each other usually are related no matter how they are sliced. 

I think an interesting experiment would be having a dictionary replacement where all strings in the starting table are replaced by random alphanumeric strings and check the performance of the encoders on that. In that case, I can imagine StringEncoder would not do so well compared to OrdinalEncoder.

rcap107 commented on doc/encoding.rst:
```diff
@@ -10,10 +10,22 @@ Encoding or vectorizing creates numerical features from the data,
-.. _dirty_categories:
+Summary
+.......
+:class:`StringEncoder` should be used in most cases when working with high-cardinality
```
StringEncoder is not rendered properly, not sure how to fix this

jeromedockes replied:
you need to add it to the reference documentation index:

https://github.com/skrub-data/skrub/blob/main/doc/reference/index.rst?plain=1#L43


so that sphinx can create a page and a link for it

rcap107 commented on skrub/_string_encoder.py:
```diff
@@ -0,0 +1,200 @@
+
+        X_out = self.vectorizer_.fit_transform(sbd.to_numpy(X))
+
+        if (min_shape := min(X_out.shape)) >= self.n_components:
+            self.tsvd_ = TruncatedSVD(n_components=self.n_components)
+            result = self.tsvd_.fit_transform(X_out)
+        else:
+            warnings.warn(
+                f"The matrix shape is {(X_out.shape)}, and its minimum is "
+                f"{min_shape}, which is too small to fit a truncated SVD with "
+                f"n_components={self.n_components}. "
+                "The embeddings will be truncated by keeping the first "
+                f"{self.n_components} dimensions instead. "
+            )
+            # self.n_components can be greater than the number
+            # of dimensions of result.
+            # Therefore, self.n_components_ below stores the resulting
+            # number of dimensions of result.
+            result = X_out[:, : self.n_components].toarray()
```
Sure, not in this PR though

jeromedockes commented on skrub/_string_encoder.py:
```diff
@@ -0,0 +1,195 @@
+
+        return result
+
+    def __sklearn_is_fitted__(self):
+        """
+        Check fitted status and return a Boolean value.
+        """
+        return check_is_fitted(self, "all_outputs_")
```
it doesn't hurt to have it but this method is not needed: by default scikit-learn's `check_is_fitted` will look for attributes with name ending with an underscore, so it would return the same result whether we override the default behavior by defining __sklearn_is_fitted__ or not

Vincent-Maladiere reviewed (commented):
OpenML downloads still fail and break the CI

jeromedockes reviewed (approved):
LGTM!! Thanks a lot @rcap107 this is a great addition :rocket:

Vincent-Maladiere reviewed (approved):
Looks good! Thanks @rcap107 :)

GaelVaroquaux commented:
I'm doing a last review of this PR. It's great, I love it. A tiny comment: I think that I prefer the comparison plot in log; if people agree, I'll implement the change
![image](https://github.com/user-attachments/assets/297ecf62-8bd7-47bd-a6d1-db4e4bd45429)

jeromedockes commented:
fine by me!

Vincent-Maladiere commented:
+1
